_posts/2023-05-26-http4s-authentication-part1.md

```diff
@@ -0,0 +1,1005 @@
+
+For this implementation, we are going to reuse our code from the Digest Authentication section.
+
+The first thing we need is a way to generate the unique code that we will send to the client.
+
+```scala
+import java.time.LocalDateTime
+import java.util.Base64
+import java.nio.charset.StandardCharsets
+
+val today: String = LocalDateTime.now().toString()
+def setToken(user: String, date: String):String = Base64.getEncoder.encodeToString(s"${user}:{$today}".getBytes(StandardCharsets.UTF_8))
+```
+
+Here we create the unique code or token using the `setToken` function which takes a username as `user` and a date computed by `LocalDateTime.now().toString()`, these values are then combined and encoded using Java's `Base64` utility.
+
+Next, we attach a new cookie to the response, this happens after a successful login.
+
+We'll also define a function that will decode the token when it's received from the client.
+
+```scala
+import scala.util.*
+
+def getUser(token: String): Try[String] = Try((new String(Base64.getDecoder.decode(token))).split(":")(0))
+```
+
+The `getUser` function will take the token and try to decode it, this can either succeed or fail as we will see later.
+
+We modify the `authedRoutes` from our Digest authentication example to add a cookie with the token information
+
+```scala
+import cats.effect.*
+import org.http4s.*
+import org.http4s.dsl.io.*
+
+case class User(id: Long, name: String)
+val authedRoutes: AuthedRoutes[User,IO] =
+    AuthedRoutes.of{
+        case GET -> Root / "welcome" as user =>
+            Ok(s"Welcome, ${user.name}").map(_.addCookie(ResponseCookie("sessioncookie", setToken(user.name, today), maxAge = Some(86400))))
+    }
+```
+
+We set our session token as a cookie on the client side using the `addCookie` function on the `Response` object. The `ResponseCookie` object takes the cookie name, `sessioncookie`, the cookie value which we set using the `setToken` function which we call with the user name and the current datetime, and finally, the maxAge, which is how long the token will persist. This is set to `Some(86400)` equal to 24 hours.
+
+Separate routes are also defined that will be accessed using our session data.
+
+```scala
+val cookieAccessRoutes = HttpRoutes.of[IO]{
+  case GET -> Root / "statement" =>
+    Ok("Financial statement processing...")
+  case GET -> Root / "logout" =>
+    Ok("Logging out...").map(_.removeCookie("sessioncookie").headers)
+}
+```
+
+The `cookieAccessRoutes` service contains two routes, `/statement` which returns financial statement information, and `/logout` which removes the `sessioncookie` token when the user is logging out. It uses the `removeCookie` function on the `Response` object.
+
+We will now need to check for the presence of our `sessioncookie` token when requests come in, here's how this can be done.
+
+```scala
+import cats.data.*
+import org.http4s.headers.Cookie
+
+def cookieCheckerService(service: HttpRoutes[IO]): HttpRoutes[IO] = Kleisli{req =>
```

Comment:
there are many nested pattern matches here, let's break it down into smaller pieces so that it's easier to read

Reply:
Broke this down and, included helper functions for each step.
Decided to use fold also to make the code more concise.